### PR TITLE
add, remove and get featured communities

### PR DESCRIPTION
**Pull request by jaswinder6991**
This PR addresses the issue https://github.com/near/neardevhub-contract/issues/36 
Adding featured-communities on the contract side.
I have added 4 functions here.
1. add_featured_community
2. get_featured_communitiesl
3. delete_featured_community 
4. is_moderator (helper function for checking if the signer is a moderator.)

I am new to this, so don't really know much about migrations. Hence I have added something from some little research I did. Let me know if it needs to be changed.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -410,6 +412,53 @@ impl Contract {
     pub fn get_community(&self, handle: String) -> Option<Community> {
         self.communities.get(&handle)
     }
+
+    pub fn add_featured_community(&mut self, handle: String) {
```
Comment (ailisp) [inline src/lib.rs]: This API would be not easy to adjust sequence of the featured communities. I recommend to set all of the featured communities: `handles: Vec<String>`.

File: src/lib.rs [hunk 2]
```diff
@@ -410,6 +412,53 @@ impl Contract {
...
+        assert!(
+            self.is_moderator(env::predecessor_account_id()),
+            "Only moderators can add featured communities"
+        );
+
+        // Check if the community already exists in the list
+        if self.featured_communities.iter().any(|fc| fc.handle == handle) {
```
Comment (ailisp) [inline src/lib.rs]: If set all, there isn't need for check already exists, but we do want to check whether every handle is really a community. You can do that by lookup from `self.communities`

**Review by ailisp — CHANGES_REQUESTED** [review]
In general code looks pretty good, but current API would be inconvenient for adjust sequence of featured community

File: src/lib.rs (hunk 1 above)
Comment (jaswinder6991) [reply src/lib.rs]: Should set all, replace the existing vector of handles or add to it at the end? I am thinking, it should remove the existing one.

Comment (jaswinder6991) [reply src/lib.rs]: Also, if this is the case, do we need remove_featured_communities function?

**Comment by akaia-shadowfox** [comment]
Closes #36

**Comment by akaia-shadowfox** [comment]
@ailisp since we'll also need to add a migration for this update, I could implement it in case no one else can tackle this.

**Comment by frol** [comment]
@near-akaia-root I will push my migration code that we produced over the call shortly. Please, don't jump on it yet.

File: src/lib.rs (hunk 1 above)
Comment (frol) [reply src/lib.rs]: I also agree that setting the whole list will be just easier to work with and cover all edit/add/remove use-cases, so let's have a single function for all of those.

P.S. In my commit https://github.com/near/neardevhub-contract/pull/42/commits/751c7bee9a16aa0a1bebd72182c3ed41995481d8, I changed `near_sdk::Vector` to `std::Vec` to store `featured_communities`.

File: src/migrations.rs [hunk 3]
```diff
@@ -132,12 +132,50 @@ impl Contract {
...
+    pub label_to_posts: UnorderedMap<String, HashSet<PostId>>,
+    pub access_control: AccessControl,
+    pub authors: UnorderedMap<AccountId, HashSet<PostId>>,
+    pub communities: UnorderedMap<String, Community>,
+}
+
+// From OldContractV3 to Contract
+#[near_bindgen]
+impl Contract {
+    fn unsafe_add_featured_communities() {
```
Comment (frol) [inline src/migrations.rs:149]: I implemented this migration over the call with @near-akaia-root explaining why migrations are necessary, and how to deal with them: https://drive.google.com/file/d/182K4BPJdHZw-TjqqTvSgnnLKBs9j7tt8/view?usp=drivesdk, and tested it over the call, but we definitely need to implement a test to check migrations (using [workspaces-rs](https://github.com/near/workspaces-rs))

File: src/lib.rs (hunk 1 above)
Comment (jaswinder6991) [reply src/lib.rs]: Thanks for the clarification, @frol. I have written the code for single function to add/remove/edit, I will pull your latest code and run some tests. Thanks for the Vector -> Vec changes, it was making a few things difficult.

File: src/migrations.rs (hunk 3 above)
Comment (jaswinder6991) [reply src/migrations.rs:149]: Thanks for the video. I will check it out too. Do we have to write migration tests in this PR? or is that a later task.

File: src/lib.rs (hunk 1 above)
Comment (jaswinder6991) [reply src/lib.rs]: Suggested changes done in the latest commit. Sample testnet widget to test - https://test.near.org/devhubopen.testnet/widget/gigs-board.pages.Feed?nearDevGovGigsContractAccountId=contract.devhubopen.testnet

File: src/lib.rs [hunk 4]
```diff
@@ -410,6 +412,40 @@ impl Contract {
     pub fn get_community(&self, handle: String) -> Option<Community> {
         self.communities.get(&handle)
     }
+
+    pub fn add_featured_communities(&mut self, handles: Vec<String>) {
```
Comment (ailisp) [inline src/lib.rs]:
```suggestion
    pub fn set_featured_communities(&mut self, handles: Vec<String>) {
```
Since it's overwrite instead of insert

**Review by ailisp — APPROVED** [review]
Looks good to me. @frol Thanks for adding migrations.

**Comment by ailisp** [comment]
@frol do we need to add migration tests in this PR?

File: src/lib.rs (hunk 4 above)
Comment (jaswinder6991) [reply src/lib.rs]: I have made the changes for the function name.

**Review by frol — APPROVED** [review]

**Comment by frol** [comment]
@ailisp I believe it should be a separate PR to address #35.

**Pull request merged**